src/main/java/org/sonar/ucfg/UCFGtoProtobuf.java

```diff
@@ -0,0 +1,151 @@
+      .setVariable(assignCall.getLhs().id())
+      .setMethodId(assignCall.getMethodId())
+      .addAllArgs(assignCall.getArgExpressions().stream().map(UCFGtoProtobuf::toProtobuf).collect(Collectors.toList()))
+      .build();
+  }
+
+  private static Ucfg.Expression toProtobuf(Expression expression) {
+    Ucfg.Expression.Builder builder = Ucfg.Expression.newBuilder();
+    if (expression.isConstant()) {
+      builder.setConst(Ucfg.Constant.newBuilder().setValue(((Expression.Constant) expression).value()).build());
+    } else {
+      builder.setVar(Ucfg.Variable.newBuilder().setName(((Expression.Variable) expression).id()).build());
+    }
+    return builder.build();
+  }
+
+  private static Ucfg.Location toProtobuf(@Nullable LocationInFile locationInFile) {
+    if (locationInFile == null) {
+      return Ucfg.Location.getDefaultInstance();
+    }
+    return Ucfg.Location.newBuilder()
+      .setFileId(locationInFile.getFileId())
+      .setStartLine(locationInFile.getStartLine())
+      .setStartLineOffset(locationInFile.getStartLineOffset())
+      .setEndLine(locationInFile.getEndLine())
+      .setEndLineOffset(locationInFile.getEndLineOffset())
+      .build();
+  }
+
+  public static UCFG fromProtobufFile(String filename) throws IOException {
+    try (FileInputStream fis = new FileInputStream(filename)) {
+      Ucfg.UCFG ucfg = Ucfg.UCFG.parseFrom(fis);
+      UCFGBuilder builder = UCFGBuilder.createUCFGForMethod(ucfg.getMethodId()).at(fromProtobuf(ucfg.getLocation()));
+      ucfg.getParametersList().forEach(pId -> builder.addMethodParam(UCFGBuilder.variableWithId(pId)));
+
+      Map<String, UCFGBuilder.BlockBuilder> blockById = ucfg.getBasicBlocksList().stream().collect(Collectors.toMap(Ucfg.BasicBlock::getId, UCFGtoProtobuf::fromProtobuf));
+      for (Map.Entry<String, UCFGBuilder.BlockBuilder> entry : blockById.entrySet()) {
+        if (ucfg.getEntriesList().contains(entry.getKey())) {
+          builder.addStartingBlock(entry.getValue());
+        } else {
+          builder.addBasicBlock(entry.getValue());
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  private static UCFGBuilder.BlockBuilder fromProtobuf(Ucfg.BasicBlock bb) {
+    UCFGBuilder.BlockBuilder blockBuilder = UCFGBuilder.newBasicBlock(bb.getId(), fromProtobuf(bb.getLocation()));
+
+    bb.getInstructionsList().forEach(
+      i -> blockBuilder.assignTo(UCFGBuilder.variableWithId(i.getVariable()),
+        UCFGBuilder.call(i.getMethodId()).withArgs(i.getArgsList().stream().map(UCFGtoProtobuf::fromProtobuf).toArray(Expression[]::new)),
+        fromProtobuf(i.getLocation())));
+
+    if(bb.hasJump()) {
+      Ucfg.Jump jump = bb.getJump();
+      blockBuilder.jumpTo(jump.getDestinationsList().stream().map(UCFGBuilder::createLabel).toArray(Label[]::new));
+    }
+    if(bb.hasRet()) {
+      Ucfg.Return ret = bb.getRet();
+      blockBuilder.ret(fromProtobuf(ret.getReturnedExpression()), fromProtobuf(ret.getLocation()));
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'ret' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.ucfg%3Asonar-ucfg&pullRequest=1&issues=AWJOLOfQ2-ApEdl5p-w4&open=AWJOLOfQ2-ApEdl5p-w4)

[ISSUE_KEY]: AWJOLOfQ2-ApEdl5p-w4